// markdown-to-docx/dist/docxBuilder.js
"use strict";
// src/docxBuilder.ts
// DOCX 문서 생성 로직
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocxBuilder = void 0;
const docx_1 = require("docx");
const fs = __importStar(require("fs"));
const inlineFormatter_1 = require("./inlineFormatter");
const imageLoader_1 = require("./imageLoader");
class DocxBuilder {
    constructor(fontFamily = "맑은 고딕", basePath) {
        this.children = [];
        this.media = new Map(); // 이미지 미디어 캐시
        this.fontFamily = fontFamily;
        this.basePath = basePath;
        this.numberingConfig = {
            config: [
                {
                    reference: "bullet-list",
                    levels: [
                        {
                            level: 0,
                            format: docx_1.LevelFormat.BULLET,
                            text: "•",
                            alignment: docx_1.AlignmentType.LEFT,
                            style: {
                                paragraph: {
                                    indent: { left: 720, hanging: 360 },
                                },
                            },
                        },
                        {
                            level: 1,
                            format: docx_1.LevelFormat.BULLET,
                            text: "◦",
                            alignment: docx_1.AlignmentType.LEFT,
                            style: {
                                paragraph: {
                                    indent: { left: 1080, hanging: 360 },
                                },
                            },
                        },
                    ],
                },
            ],
        };
    }
    addElement(element) {
        if (element.type === "heading") {
            this.addHeading(element);
        }
        else if (element.type === "table") {
            this.addTable(element);
        }
        else if (element.type === "blockquote") {
            this.addBlockquote(element);
        }
        else if (element.type === "checkbox") {
            this.addCheckbox(element);
        }
        else if (element.type === "list") {
            this.addList(element);
        }
        else if (element.type === "image") {
            // 이미지 로드는 비동기이므로 Promise를 반환하지만,
            // addElement는 동기이므로 실행 후 결과를 await하지 않음
            // catch 처리로 에러 처리
            this.addImage(element).catch((err) => {
                // 이미지 로드 실패 시, 에러 메시지로 대체
                const errorMsg = err instanceof Error ? err.message : "Unknown error";
                const errorText = `[Image Error: ${errorMsg}]`;
                this.addParagraph({ type: "paragraph", text: errorText });
            });
        }
        else if (element.type === "paragraph") {
            this.addParagraph(element);
        }
    }
    addHeading(element) {
        const headingLevelMap = {
            1: docx_1.HeadingLevel.HEADING_1,
            2: docx_1.HeadingLevel.HEADING_2,
            3: docx_1.HeadingLevel.HEADING_3,
            4: docx_1.HeadingLevel.HEADING_4,
            5: docx_1.HeadingLevel.HEADING_5,
            6: docx_1.HeadingLevel.HEADING_6,
        };
        const headingLevel = headingLevelMap[element.level] ?? docx_1.HeadingLevel.HEADING_1;
        this.children.push(new docx_1.Paragraph({
            heading: headingLevel,
            children: (0, inlineFormatter_1.parseInlineFormatting)(element.text, this.fontFamily),
        }));
    }
    addTable(element) {
        const headerCells = element.headers.map((header) => new docx_1.TableCell({
            children: [
                new docx_1.Paragraph({
                    children: (0, inlineFormatter_1.parseInlineFormatting)(header, this.fontFamily),
                }),
            ],
            shading: { fill: "D3D3D3" },
            verticalAlign: docx_1.VerticalAlign.CENTER,
        }));
        const bodyRows = element.rows.map((row) => new docx_1.TableRow({
            children: row.map((cell) => new docx_1.TableCell({
                children: [
                    new docx_1.Paragraph({
                        children: (0, inlineFormatter_1.parseInlineFormatting)(cell, this.fontFamily),
                    }),
                ],
            })),
        }));
        this.children.push(new docx_1.Table({
            width: { size: 100, type: "pct" },
            rows: [
                new docx_1.TableRow({
                    children: headerCells,
                }),
                ...bodyRows,
            ],
            borders: {
                top: { style: docx_1.BorderStyle.SINGLE, size: 1, color: "000000" },
                bottom: { style: docx_1.BorderStyle.SINGLE, size: 1, color: "000000" },
                left: { style: docx_1.BorderStyle.SINGLE, size: 1, color: "000000" },
                right: { style: docx_1.BorderStyle.SINGLE, size: 1, color: "000000" },
                insideHorizontal: {
                    style: docx_1.BorderStyle.SINGLE,
                    size: 1,
                    color: "000000",
                },
                insideVertical: {
                    style: docx_1.BorderStyle.SINGLE,
                    size: 1,
                    color: "000000",
                },
            },
        }));
        this.children.push(new docx_1.Paragraph({
            text: "",
            spacing: { after: 200 },
        }));
    }
    addBlockquote(element) {
        this.children.push(new docx_1.Paragraph({
            children: (0, inlineFormatter_1.parseInlineFormatting)(element.text, this.fontFamily),
            indent: { left: 360 },
            spacing: { before: 200, after: 200 },
            shading: {
                fill: "F6F8FA",
            },
            border: {
                left: {
                    color: "0969DA",
                    space: 4,
                    style: docx_1.BorderStyle.SINGLE,
                    size: 24,
                },
            },
        }));
    }
    addCheckbox(element) {
        const checkmark = element.checked ? "☑" : "☐";
        this.children.push(new docx_1.Paragraph({
            children: [
                new docx_1.TextRun({
                    text: checkmark + " ",
                    font: this.fontFamily,
                    size: 22,
                }),
                ...(0, inlineFormatter_1.parseInlineFormatting)(element.text, this.fontFamily),
            ],
            numbering: {
                reference: "bullet-list",
                level: element.level,
            },
            spacing: { before: 40, after: 80 },
        }));
    }
    addList(element) {
        this.children.push(new docx_1.Paragraph({
            children: (0, inlineFormatter_1.parseInlineFormatting)(element.text, this.fontFamily),
            numbering: {
                reference: "bullet-list",
                level: element.level,
            },
            spacing: { before: 40, after: 80 },
        }));
    }
    addParagraph(element) {
        const paragraphLines = element.text.split("\n");
        paragraphLines.forEach((line, index) => {
            if (line.trim() || index > 0) {
                this.children.push(new docx_1.Paragraph({
                    children: (0, inlineFormatter_1.parseInlineFormatting)(line, this.fontFamily),
                    spacing: { after: 200 },
                }));
            }
        });
    }
    build() {
        return new docx_1.Document({
            styles: {
                default: {
                    document: {
                        run: {
                            font: this.fontFamily,
                            size: 22,
                        },
                        paragraph: {
                            spacing: { line: 276, lineRule: "auto" },
                        },
                    },
                },
                paragraphStyles: [
                    {
                        id: "Heading1",
                        name: "Heading 1",
                        basedOn: "Normal",
                        next: "Normal",
                        quickFormat: true,
                        run: {
                            font: this.fontFamily,
                            size: 56,
                            bold: true,
                            color: "000000",
                        },
                        paragraph: {
                            spacing: { before: 400, after: 320 },
                        },
                    },
                    {
                        id: "Heading2",
                        name: "Heading 2",
                        basedOn: "Normal",
                        next: "Normal",
                        quickFormat: true,
                        run: {
                            font: this.fontFamily,
                            size: 42,
                            bold: true,
                            color: "000000",
                        },
                        paragraph: {
                            spacing: { before: 380, after: 200 },
                        },
                    },
                    {
                        id: "Heading3",
                        name: "Heading 3",
                        basedOn: "Normal",
                        next: "Normal",
                        quickFormat: true,
                        run: {
                            font: this.fontFamily,
                            size: 32,
                            bold: true,
                            color: "000000",
                        },
                        paragraph: {
                            spacing: { before: 360, after: 160 },
                        },
                    },
                    {
                        id: "Heading4",
                        name: "Heading 4",
                        basedOn: "Normal",
                        next: "Normal",
                        quickFormat: true,
                        run: {
                            font: this.fontFamily,
                            size: 28,
                            bold: true,
                            color: "000000",
                        },
                        paragraph: {
                            spacing: { before: 320, after: 120 },
                        },
                    },
                ],
            },
            numbering: this.numberingConfig,
            sections: [
                {
                    properties: {
                        page: {
                            margin: {
                                top: 1440,
                                right: 1440,
                                bottom: 1440,
                                left: 1440,
                            },
                        },
                    },
                    children: this.children,
                },
            ],
        });
    }
    async addImage(element) {
        const imageResult = await (0, imageLoader_1.loadImage)(element.src, this.basePath);
        if (!imageResult.success || imageResult.buffer.length === 0) {
            // 이미지 로드 실패 시 alt 텍스트로 대체
            const errorMsg = imageResult.error || "Failed to load image";
            this.addParagraph({
                type: "paragraph",
                text: `[Image: ${element.alt}] (${errorMsg})`,
            });
            return;
        }
        // 이미지 캐시에 저장 (중복 방지)
        const cacheKey = `${element.src}`;
        if (!this.media.has(cacheKey)) {
            this.media.set(cacheKey, imageResult.buffer);
        }
        // 이미지를 포함한 단락 생성
        // docx 라이브러리는 직접 이미지 객체를 지원하지 않으므로,
        // 현재는 이미지 메타데이터를 주석으로 표현
        // 향후 docx 라이브러리 업데이트 시 실제 이미지 삽입 가능
        const paragraph = new docx_1.Paragraph({
            children: [
                new docx_1.TextRun({
                    text: `[Image: ${element.alt}]`,
                    italics: true,
                    color: "808080",
                }),
            ],
            spacing: { before: 240, after: 240 },
        });
        this.children.push(paragraph);
        console.log(`Added image: ${element.alt} (${element.src})`);
    }
    async save(outputFile) {
        const doc = this.build();
        const buffer = await docx_1.Packer.toBuffer(doc);
        fs.writeFileSync(outputFile, buffer);
        return outputFile;
    }
}
exports.DocxBuilder = DocxBuilder;
