// markdown-to-docx/dist/imageLoader.js
"use strict";
// src/imageLoader.ts
// 이미지 로딩 및 변환 유틸리티
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadLocalImage = loadLocalImage;
exports.loadRemoteImage = loadRemoteImage;
exports.loadImage = loadImage;
exports.getDocxMediaExtension = getDocxMediaExtension;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
/**
 * 파일 경로에서 MIME 타입 추론
 */
function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const mimeTypes = {
        ".png": "image/png",
        ".jpg": "image/jpeg",
        ".jpeg": "image/jpeg",
        ".gif": "image/gif",
        ".bmp": "image/bmp",
        ".webp": "image/webp",
    };
    return mimeTypes[ext] || "image/png";
}
/**
 * URL이 HTTP/HTTPS인지 확인
 */
function isHttpUrl(src) {
    try {
        const url = new URL(src);
        return url.protocol === "http:" || url.protocol === "https:";
    }
    catch {
        return false;
    }
}
/**
 * 로컬 파일에서 이미지 로드
 */
function loadLocalImage(filePath, basePath) {
    try {
        // 상대 경로인 경우 basePath와 결합
        let resolvedPath = filePath;
        if (basePath && !path.isAbsolute(filePath)) {
            resolvedPath = path.join(basePath, filePath);
        }
        // 파일 존재 확인
        if (!fs.existsSync(resolvedPath)) {
            return {
                buffer: Buffer.alloc(0),
                mimeType: "image/png",
                success: false,
                error: `File not found: ${resolvedPath}`,
            };
        }
        // 파일 읽기
        const buffer = fs.readFileSync(resolvedPath);
        const mimeType = getMimeType(resolvedPath);
        return {
            buffer,
            mimeType,
            success: true,
        };
    }
    catch (err) {
        const errorMsg = err instanceof Error ? err.message : "Unknown error";
        return {
            buffer: Buffer.alloc(0),
            mimeType: "image/png",
            success: false,
            error: `Failed to load image: ${errorMsg}`,
        };
    }
}
/**
 * HTTP(S) URL에서 이미지 다운로드 (Node.js built-in fetch 또는 https/http 모듈)
 * Node.js 18.x+ 사용 시 fetch 지원
 */
async function loadRemoteImage(url) {
    try {
        // Node.js 18+ fetch 사용
        if (typeof fetch !== "undefined") {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), 10000);
            try {
                const response = await fetch(url, {
                    signal: controller.signal,
                });
                clearTimeout(timeoutId);
                if (!response.ok) {
                    return {
                        buffer: Buffer.alloc(0),
                        mimeType: "image/png",
                        success: false,
                        error: `HTTP ${response.status}: ${response.statusText}`,
                    };
                }
                const arrayBuffer = await response.arrayBuffer();
                const buffer = Buffer.from(arrayBuffer);
                const contentType = response.headers.get("content-type") || "image/png";
                const mimeType = contentType.split(";")[0].trim();
                return {
                    buffer,
                    mimeType,
                    success: true,
                };
            }
            catch (fetchErr) {
                clearTimeout(timeoutId);
                throw fetchErr;
            }
        }
        // 폴백: https/http 모듈 사용 (Node.js < 18)
        return await loadRemoteImageFallback(url);
    }
    catch (err) {
        const errorMsg = err instanceof Error ? err.message : "Unknown error";
        return {
            buffer: Buffer.alloc(0),
            mimeType: "image/png",
            success: false,
            error: `Failed to download image: ${errorMsg}`,
        };
    }
}
// CommonJS require 모듈 (Node.js http/https 서버 요청용)
// import 대신 require 사용: Node.js 내장 모듈 CommonJS 호환성
const https = require("https");
const http = require("http");
/**
 * https/http 모듈을 사용한 폴백 다운로드
 */
async function loadRemoteImageFallback(urlString) {
    return new Promise((resolve) => {
        const url = new URL(urlString);
        const isHttps = url.protocol === "https:";
        const httpModule = isHttps ? https : http;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const request = httpModule.get(urlString, { timeout: 10000 }, (response) => {
            const chunks = [];
            if (response.statusCode !== 200) {
                resolve({
                    buffer: Buffer.alloc(0),
                    mimeType: "image/png",
                    success: false,
                    error: `HTTP ${response.statusCode}`,
                });
                response.resume(); // consume response to free up memory
                return;
            }
            response.on("data", (chunk) => {
                chunks.push(chunk);
            });
            response.on("end", () => {
                const buffer = Buffer.concat(chunks);
                const contentType = response.headers["content-type"] || "image/png";
                const mimeType = contentType.split(";")[0].trim();
                resolve({
                    buffer,
                    mimeType,
                    success: true,
                });
            });
            response.on("error", (err) => {
                resolve({
                    buffer: Buffer.alloc(0),
                    mimeType: "image/png",
                    success: false,
                    error: `Download error: ${err.message}`,
                });
            });
        });
        request.on("error", (err) => {
            resolve({
                buffer: Buffer.alloc(0),
                mimeType: "image/png",
                success: false,
                error: `Request error: ${err.message}`,
            });
        });
        request.on("timeout", () => {
            request.destroy();
            resolve({
                buffer: Buffer.alloc(0),
                mimeType: "image/png",
                success: false,
                error: "Request timeout (10s)",
            });
        });
    });
}
/**
 * 이미지 소스(경로 또는 URL)에서 이미지 로드 (동기/비동기 래퍼)
 */
async function loadImage(src, basePath) {
    if (isHttpUrl(src)) {
        return loadRemoteImage(src);
    }
    // 로컬 파일은 동기 로드
    return loadLocalImage(src, basePath);
}
/**
 * MIME 타입에서 docx Media 확장자 추론
 */
function getDocxMediaExtension(mimeType) {
    const extensions = {
        "image/png": "png",
        "image/jpeg": "jpg",
        "image/gif": "gif",
        "image/bmp": "bmp",
        "image/webp": "webp",
    };
    return extensions[mimeType] || "png";
}
